hibernate-core/src/main/java/org/hibernate/collection/spi/AbstractPersistentCollection.java

```diff
@@ -282,8 +283,8 @@ else if ( !session.isConnected() ) {
 				}
 			}
 			else {
-				if ( !session.isTransactionInProgress() ) {
-					session.getJdbcCoordinator().afterTransaction();
+				if ( !session.isTransactionInProgress() && !isPhysicallyConnected ) {
+					session.getJdbcCoordinator().getLogicalConnection().afterTransaction();
```

Comment:
Can you add a comment here about why we only have to call `afterTransaction` if this is physically connected? Also, I'm having a hard time understanding why you changed from `session.getJdbcCoordinator().afterTransaction();` to `session.getJdbcCoordinator().getLogicalConnection().afterTransaction();` as the former code seemed to invoke some other code as well, see `org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl#afterTransaction`.

Overall, I am simply not sure I understand how this change is necessary for the subselect fetch issue and I think a code comment might help here.

Reply:
`session.getJdbcCoordinator().afterTransaction()` it is fine, I can revert this.

About the check to `isPhysicallyConnected`, not able to find a better way. The issue is that an `SelectEagerCollectionInitializer#initilaizeInstance` calls `persistenceContext.initializeNonlazyCollections()` that triggers this code that closes the connection but then when we iterate over the others initializers calling initializeInstance() they will not able to iterate the ResultSet. 
Hope my explanation make sense :)